sonic-slave-buster/Dockerfile.j2

```diff
@@ -40,11 +48,88 @@ RUN echo "deb [arch=arm64] http://deb.debian.org/debian buster main contrib non-
+ARG arch=arm64
+ARG CROSS_CC=aarch64-linux-gnu-gcc
+ARG CROSS_CXX=aarch64-linux-gnu-g++
+ARG PYTHON_CROSS_PLATFORM=linux_aarch64
 {%- endif %}
 
 ## Make apt-get non-interactive
 ENV DEBIAN_FRONTEND=noninteractive
 
+{%- if CROSS_BUILD_ENVIRON == "y" %}
+RUN apt-get update && apt-get install -y python3 python3-pip
+
+RUN apt-get update && apt-get install -y \
+	python:$arch python3-minimal:$arch python3.7:$arch python3:$arch python-setuptools:$arch 
+RUN apt-get download python3-distutils && dpkg --force-all -i python3-distutils*
+RUN apt-get download python3-pip && dpkg --force-all -i python3-pip*
+RUN which pip3 && pip3 install enum34
+RUN pip3 install virtualenv
+
+RUN file /usr/bin/python2.7 && file /usr/bin/python3.7
+
+# Create target arm python2/3 virtual environments with all required packages installed
+RUN mkdir /python_virtualenv
+RUN cd /python_virtualenv && python3 -m virtualenv -p /usr/bin/python env2
+RUN cd /python_virtualenv && python3 -m virtualenv --copies -p /usr/bin/python3 env3
+
+RUN PATH=/python_virtualenv/env2/bin/:$PATH pip2 install setuptools==40.8.0 wheel==0.35.1 fastentrypoints pytest pytest-cov pytest-runner==4.4 nose==1.3.7 mockredispy==2.9.3 mock==3.0.5 j2cli==0.3.10 PyYAML==5.4.1 pexpect==4.6.0 Pympler==0.8 ctypesgen==1.0.2 natsort redis
+RUN PATH=/python_virtualenv/env3/bin/:$PATH pip3 install setuptools==49.6.00 wheel==0.35.1 fastentrypoints pytest pytest-cov pytest-runner==5.2 nose==1.3.7 mockredispy==2.9.3 mock==3.0.5 PyYAML==5.4.1 redis pexpect==4.8.0 Pympler==0.8 parameterized natsort Jinja2==2.10 
+
+RUN apt-get --fix-broken install -y
+RUN apt-get install -y libpython2.7-dev:$arch libpython3.7-dev:$arch libxml2-dev:$arch libxslt-dev:$arch libssl-dev:$arch libz-dev:$arch
+
+RUN file /python_virtualenv/env3/bin/python3 && file /python_virtualenv/env2/bin/python
+
+RUN apt-get install -y swig libssl-dev
+
+COPY ./cross-build-arm-python-reqirements.sh ./
+# Cross-compile the heaviest python packages for arm
+ARG pre_built_target_dir=PRE_BUILT_TARGET 
+RUN PATH=/python_virtualenv/env2/bin/:/python_virtualenv/env3/bin/:$PATH CC=$CROSS_CC PY_PLAT=$PYTHON_CROSS_PLATFORM ./cross-build-arm-python-reqirements.sh $pre_built_target_dir && ls -l $pre_built_target_dir
+RUN rm ./cross-build-arm-python-reqirements.sh
+RUN PATH=/python_virtualenv/env2/bin/:$PATH pip install $pre_built_target_dir/*cp27*.whl
+RUN PATH=/python_virtualenv/env3/bin/:$PATH pip3 install $pre_built_target_dir/*cp37*.whl
+
+RUN PATH=/python_virtualenv/env3/bin/:$PATH python3 -m pip list
+
+# Depend on cross-compiled lxml
+RUN PATH=/python_virtualenv/env3/bin/:$PATH python3 -m pip install pyang==2.4.0
+RUN PATH=/python_virtualenv/env3/bin/:$PATH python3 -m pip install pyangbind==0.8.1
+RUN PATH=/python_virtualenv/env3/bin/:$PATH python3 -m pip uninstall -y enum34
+RUN PATH=/python_virtualenv/env3/bin/:$PATH pip3 install --force-reinstall --no-cache-dir coverage
+{%- endif %}
+
 RUN apt-get update && apt-get install -y \
```

Comment:
In reference to comment given above: 

> Here is a sample from the Dockerfile.j2:
> RUN apt-get update && apt-get install -y libelf-dev:$arch libdw-dev:$arch libbz2-dev:$arch liblzo2-dev:$arch libedit-dev:$arch libevent-dev:$arch libopts25-dev:$arch ...
> where $arch is either arm64 or armhf according to the target architecture.

In that case, shoudn't we be adding :arch extensions to those packages which are target specific. 

Because otherwise, we might face these problems in the future.
For Eg: I am working to compile a package which depend on "libusb-1.0-0-dev:arm64" and we already install this package here 

```
# For BFN sdk build
        libusb-1.0-0-dev \ (host arch package will be fetched here and thus it won't work for cross compilation scenario )
```

But also, packages like these need not be target specific, they have to be based on host arch

```
# For build image
        cpio \
        squashfs-tools \
        zip \
```

Any idea/plans on solving this?

Reply:
Usually there is no problem in installing the same package for both host and other architectures. They are installed in different directories. You can see in Dockerfile.j2 that some packages are already installed for both host and armhf/arm64 architectures.